File: go/internal/prompts/prompts.go
package prompts

import (
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"sync"
)

var (
	dirOnce sync.Once
	dirVal  string
	dirErr  error

	cacheMu  sync.Mutex
	tplCache = map[string]string{}
)

func Dir() (string, error) {
	dirOnce.Do(func() {
		var attempted []string
		if v := os.Getenv("CODEX_WEB_PROMPTS"); v != "" {
			attempted = append(attempted, v)
			if isPromptsDir(v) {
				dirVal = v
				return
			}
		}
		if exe, err := os.Executable(); err == nil {
			if p := walkUpForPrompts(filepath.Dir(exe), &attempted); p != "" {
				dirVal = p
				return
			}
		}
		if cwd, err := os.Getwd(); err == nil {
			if p := walkUpForPrompts(cwd, &attempted); p != "" {
				dirVal = p
				return
			}
		}
		dirErr = fmt.Errorf("prompts directory not found; tried: %s", strings.Join(attempted, ", "))
	})
	return dirVal, dirErr
}

func walkUpForPrompts(start string, attempted *[]string) string {
	cur := start
	for i := 0; i < 10; i++ {
		candidate := filepath.Join(cur, "prompts")
		*attempted = append(*attempted, candidate)
		if isPromptsDir(candidate) {
			return candidate
		}
		parent := filepath.Dir(cur)
		if parent == cur {
			break
		}
		cur = parent
	}
	return ""
}

func isPromptsDir(p string) bool {
	info, err := os.Stat(p)
	return err == nil && info.IsDir()
}

func Render(toolName string, vars map[string]string) (string, error) {
	cacheMu.Lock()
	tpl, ok := tplCache[toolName]
	cacheMu.Unlock()
	if !ok {
		d, err := Dir()
		if err != nil {
			return "", err
		}
		data, err := os.ReadFile(filepath.Join(d, toolName+".md"))
		if err != nil {
			return "", fmt.Errorf("read prompt %s: %w", toolName, err)
		}
		tpl = string(data)
		cacheMu.Lock()
		tplCache[toolName] = tpl
		cacheMu.Unlock()
	}
	out := tpl
	for k, v := range vars {
		out = strings.ReplaceAll(out, "{"+k+"}", v)
	}
	return out, nil
}

File: python/src/codex_web_mcp/prompts.py
"""Prompt template loader. Resolution: env var -> walk up from executable -> walk up from cwd."""

from __future__ import annotations

import os
import sys
import threading
from pathlib import Path

_dir_lock = threading.Lock()
_dir_value: Path | None = None
_tpl_lock = threading.Lock()
_tpl_cache: dict[str, str] = {}


def _is_prompts_dir(p: Path) -> bool:
    return p.is_dir()


def _walk_up(start: Path, attempted: list[str]) -> Path | None:
    cur = start
    for _ in range(10):
        cand = cur / "prompts"
        attempted.append(str(cand))
        if _is_prompts_dir(cand):
            return cand
        parent = cur.parent
        if parent == cur:
            break
        cur = parent
    return None


def resolve_dir() -> Path:
    global _dir_value
    with _dir_lock:
        if _dir_value is not None:
            return _dir_value
        attempted: list[str] = []
        env = os.environ.get("CODEX_WEB_PROMPTS")
        if env:
            attempted.append(env)
            ep = Path(env)
            if _is_prompts_dir(ep):
                _dir_value = ep
                return _dir_value
        try:
            exe_dir = Path(sys.argv[0]).resolve().parent
            found = _walk_up(exe_dir, attempted)
            if found:
                _dir_value = found
                return _dir_value
        except OSError:
            pass
        try:
            cwd = Path.cwd()
            found = _walk_up(cwd, attempted)
            if found:
                _dir_value = found
                return _dir_value
        except OSError:
            pass
        raise FileNotFoundError(
            "prompts directory not found; tried: " + ", ".join(attempted)
        )


def render(tool_name: str, vars: dict[str, str]) -> str:
    with _tpl_lock:
        tpl = _tpl_cache.get(tool_name)
    if tpl is None:
        d = resolve_dir()
        path = d / f"{tool_name}.md"
        tpl = path.read_text(encoding="utf-8")
        with _tpl_lock:
            _tpl_cache[tool_name] = tpl
    out = tpl
    for k, v in vars.items():
        out = out.replace("{" + k + "}", v)
    return out
